pyccl/halos/halo_model_base.py

```diff
@@ -0,0 +1,300 @@
+from .. import ccllib as lib
+from ..core import check
+from ..parameters import physical_constants as const
+from ..base import CCLAutoreprObject, warn_api, deprecated, deprecate_attr
+import numpy as np
+import functools
+from abc import abstractmethod
+
+
+__all__ = ("HMIngredients", "TinkerFunction",)
+
+
+def _subclasses(cls):
+    # This helper returns a set of all subclasses.
+    direct_subs = cls.__subclasses__()
+    deep_subs = [sub for cl in direct_subs for sub in cl._subclasses()]
+    return set(direct_subs).union(deep_subs)
+
+
+def from_name(cls, name):
+    """Obtain particular model."""
+    mod = {p.name: p for p in cls._subclasses() if hasattr(p, "name")}
+    return mod[name]
+
+
+def initialize_from_input(cls, input_, **kwargs):
+    """Process the input and generate an object of the class.
+    Input can be an instance of the class, or a name string.
+    Optional ``**kwargs`` may be passed.
+    """
+    if isinstance(input_, cls):
+        return input_
+    if isinstance(input_, str):
+        class_ = cls.from_name(input_)
+        return class_(**kwargs)
+    good, bad = cls.__name__, input_.__class__.__name__
+    raise TypeError(f"Expected {good} or str but received {bad}.")
+
+
+class HMIngredients(CCLAutoreprObject):
+    """Base class for halo model ingredients."""
+    __repr_attrs__ = ("mass_def", "mass_def_strict",)
+    __getattr__ = deprecate_attr(pairs=[('mdef', 'mass_def')]
+                                 )(super.__getattribute__)
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        cls._subclasses = classmethod(_subclasses)
+        cls.from_name = classmethod(from_name)
+        cls.initialize_from_input = classmethod(initialize_from_input)
+
+    @warn_api
+    def __init__(self, *, mass_def, mass_def_strict=True):
+        self.mass_def_strict = mass_def_strict
+        # Check if mass definition was provided and check that it's sensible.
+        if self._check_mass_def(mass_def):
+            classname = self.__class__.name
+            raise ValueError(
+                f"{classname} is not defined for {mass_def.name}-based masses."
+                " To disable this exception set `mass_def_strict=True`.")
+        self.mass_def = mass_def
+        self._setup()
+
+    @abstractmethod
+    def _check_mass_def_strict(self, mass_def):
+        """Check if this class is defined for mass definition ``mass_def``."""
+
+    def _setup(self):
```

Comment:
No because users may or may not implement one. If something is an abstract method then it **must** have an implementation or the class won't instantiate.

Reply:
FYI: Internally, `ABCMeta`, the metaclass which controls `ABC` in `abc`, defines an `__abstractmethods__` hook in all of its subclasses. Before `__init__` is run, but after `__new__` is run, it checks whether the class contains any abstract methods. If it does, it raises.